src/msw/srchctrl.cpp

```diff
@@ -0,0 +1,448 @@
+                               WXWPARAM wParam,
+                               WXLPARAM lParam)
+{
+    bool processed = false;
+    const int bitmapWidth = GetSize().GetHeight();
+
+    switch ( nMsg )
+    {
+        case WM_NCCALCSIZE:
+            {
+                if ( wParam )
+                {
+                    NCCALCSIZE_PARAMS* csparam = (NCCALCSIZE_PARAMS *)lParam;
+
+                    // Where to draw the buttons depending on the layout,
+                    //    LTR: search button (left), cancel button (right).
+                    //    RTL: search button (right), cancel button (left).
+
+                    if ( GetLayoutDirection() == wxLayout_LeftToRight )
+                    {
+                        csparam->rgrc [0].left  += bitmapWidth - 2;
+
+                        if ( IsCancelButtonVisible() )
+                        {
+                            csparam->rgrc [0].right -= bitmapWidth - 2;
+                        }
+                    }
+                    else // wxLayout_RightToLeft
+                    {
+                        csparam->rgrc [0].right -= bitmapWidth - 2;
+
+                        if ( IsCancelButtonVisible() )
+                        {
+                            csparam->rgrc [0].left  += bitmapWidth - 2;
+                        }
+                    }
+                }
+            }
+            break;
+
+        case WM_NCPAINT:
+            {
+                const RECT rcWindow = wxGetWindowRect(GetHwnd());
+                
+                RECT rcClient = wxGetClientRect(GetHwnd());
+                ::MapWindowPoints(GetHwnd(), NULL, reinterpret_cast<LPPOINT>(&rcClient), 2);
+
+                RECT leftRect, rightRect;
+                RECT *searchButtonRect, *cancelButtonRect;
+
+                if ( GetLayoutDirection() == wxLayout_LeftToRight )
+                {
+                    searchButtonRect = &leftRect;
+                    cancelButtonRect = &rightRect;
+                }
+                else // wxLayout_RightToLeft
+                {
+                    searchButtonRect = &rightRect;
+                    cancelButtonRect = &leftRect;
+                }
+
+                // In RTL layout, we can skip leftRect (cancel button) calculation.
+                if ( !(GetLayoutDirection() == wxLayout_RightToLeft && !IsCancelButtonVisible()) )
+                {
+                    // Calculate leftRect
+                    leftRect = rcWindow;
+                    leftRect.right = leftRect.left + bitmapWidth;
+
+                    leftRect.top += rcClient.top - rcWindow.top;
+                    leftRect.bottom -= rcWindow.bottom - rcClient.bottom;
+
+                    ::OffsetRect(&leftRect, -rcWindow.left, -rcWindow.top);
+                }
+                
+                // In LTR layout, we can skip rightRect (cancel button) calculation.
+                if ( !(GetLayoutDirection() == wxLayout_LeftToRight && !IsCancelButtonVisible()) )
```

Comment:
why double negatives? `GetLayoutDirection() == wxLayout_LeftToRight && IsCancelButtonVisible()` should do the same.